pkg/server/api/svid/v1/service.go

```diff
@@ -324,6 +328,96 @@ func (s *Service) NewJWTSVID(ctx context.Context, req *svid.NewJWTSVIDRequest) (
 	}, nil
 }
 
-func (s *Service) NewDownstreamX509CA(context.Context, *svid.NewDownstreamX509CARequest) (*svid.NewDownstreamX509CAResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "not implemented")
+func (s *Service) newX509CASVID(ctx context.Context, entry *types.Entry, csrBytes []byte) (resp *svid.NewDownstreamX509CAResponse, err error) {
+	log := rpccontext.Logger(ctx)
+
+	if !entry.Downstream {
+		err := status.Errorf(codes.NotFound, "error not found or not authorized")
+		log.WithError(err).Error("Rejecting NewDownstreamX509CA request due to downstream flag not being set")
+		return nil, err
+	}
+
+	spiffeID, err := api.IDFromProto(entry.SpiffeId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "malformed entry")
+	}
+
+	csr, err := x509.ParseCertificateRequest(csrBytes)
+	if err != nil {
+		log.WithError(err).Error("Invalid request: malformed CSR")
+		return nil, status.Errorf(codes.InvalidArgument, "malformed CSR: %v", err)
+	}
+
+	if err := csr.CheckSignature(); err != nil {
+		log.WithError(err).Error("Invalid request: invalid CSR signature")
+		return nil, status.Error(codes.InvalidArgument, "invalid CSR signature")
+	}
+
+	x509CASvid, err := s.ca.SignX509CASVID(ctx, ca.X509CASVIDParams{
+		SpiffeID:  spiffeID.String(),
+		PublicKey: csr.PublicKey,
+		TTL:       time.Duration(entry.Ttl) * time.Second,
+	})
+	if err != nil {
+		log.WithError(err).Error("Failed to sign X509-SVID")
+		return nil, status.Errorf(codes.Internal, "failed to sign X509-SVID: %v", err)
+	}
+
+	dsResp, err := s.ds.FetchBundle(ctx, &datastore.FetchBundleRequest{
+		TrustDomainId: s.td.IDString(),
+	})
+	if err != nil {
+		log.Errorf("Failed to fetch bundle: %v", err)
+		return nil, status.Errorf(codes.Internal, "failed to fetch bundle: %v", err)
+	}
+
+	if dsResp.Bundle == nil {
+		log.Error("Bundle not found")
+		return nil, status.Error(codes.Internal, "bundle not found")
+	}
+
+	rawRootCerts := make([][]byte, 0, len(dsResp.Bundle.RootCas))
+	for _, cert := range dsResp.Bundle.RootCas {
+		rawRootCerts = append(rawRootCerts, cert.DerBytes)
+	}
+
+	return &svid.NewDownstreamX509CAResponse{
+		CaCertChain:     x509util.RawCertsFromCertificates(x509CASvid),
+		X509Authorities: rawRootCerts,
+	}, nil
+}
+
+func (s *Service) NewDownstreamX509CA(ctx context.Context, req *svid.NewDownstreamX509CARequest) (*svid.NewDownstreamX509CAResponse, error) {
+	log := rpccontext.Logger(ctx)
+
+	if err := rpccontext.RateLimit(ctx, 1); err != nil {
+		log.WithError(err).Error("Rejecting request due to downstream CA signing rate limit")
+		return nil, err
+	}
+
+	entriesMap, err := s.fetchEntries(ctx, log)
+	if err != nil {
+		return nil, err
+	}
+
+	entry, ok := entriesMap[req.EntryId]
+	if !ok {
+		err := status.Errorf(codes.NotFound, "error not found or not authorized")
+		log.WithError(err).Error("Rejecting NewDownstreamX509CA request due to invalid entry ID")
+		return nil, err
+	}
```

Comment:
We may have a helper function for this. This is also done in `NewJWTSVID` RPC.

Reply:
I don't think a shared function would save any code here since it's such a simple check (still, can do it if you think)

Reply:
Just to make sure you guys saw my comment on this section, but this code block isn't the right thing for this RPC (since it isn't doing an authorized entry crawl for an agent).

Reply:
Just saw it. Thanks!